notebook-api/notebook/notebook.py

```diff
@@ -0,0 +1,434 @@
+                'uuid': pod_id,
+                **labels}),
+        spec=service_spec)
+    svc = k8s.create_namespaced_service(
+        'default',
+        service_template,
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS
+    )
+    pod_spec = kube.client.V1PodSpec(
+        containers=[
+            kube.client.V1Container(
+                command=[
+                    'jupyter',
+                    'notebook',
+                    f'--NotebookApp.token={jupyter_token}',
+                    f'--NotebookApp.base_url=/instance/{svc.metadata.name}/'
+                ],
+                name='default',
+                image=image,
+                ports=[kube.client.V1ContainerPort(container_port=8888)],
+                resources=kube.client.V1ResourceRequirements(
+                     requests={'cpu': '1.601', 'memory': '1.601G'}),
+                readiness_probe=kube.client.V1Probe(
+                    http_get=kube.client.V1HTTPGetAction(
+                        path=f'/instance/{svc.metadata.name}/login',
+                        port=8888)))])
+    pod_template = kube.client.V1Pod(
+        metadata=kube.client.V1ObjectMeta(
+            generate_name='notebook-worker-',
+            labels={
+                'app': 'notebook-worker',
+                'hail.is/notebook-instance': INSTANCE_ID,
+                'uuid': pod_id,
+                'svc_name': svc.metadata.name,
+                **labels
+            },),
+        spec=pod_spec)
+    pod = k8s.create_namespaced_pod(
+        'default',
+        pod_template,
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS,
+    )
+
+    # Avoid extra queries to get pod_name (or reciprocol, svc_name)
+    # svc.metadata.labels['pod_name'] = pod.metadata.name
+    # k8s.patch_namespaced_service(svc.metadata.name, 'default', {
+    #     'metadata': {
+    #         'labels': {
+    #             'pod_name': pod.metadata.name
+    #         }
+    #     }
+    # })
+    return svc, pod
+
+
+def forbidden():
+    return 'Forbidden', 403
+
+# Have consumer pass a graph to awalk
+
+# TODO: could support array values as well, i.e path.to.0
+
+
+def get_path(data, path: str):
+    idx = path.find('.')
+    if idx == -1:
+        if isinstance(data, dict):
+            return data[path]
+        return getattr(data, path)
+
+    return get_path(getattr(data, path[0:idx]), path[idx + 1:])
+
+# Could make this somewhat more elegenat by checking whether resources is list or object
+# if object, return json object, else json array; this may come at 1 additional functional call per resource
+# or an uglier inline block
+# Given a kubernetes object, and some collection of field paths, walk path tree and fetch values
+# @param resources<List> : kubernetes v1 object
+# @param paths<List<List[3]>> : fields and transformations: [key, path_in_kube_object, lambda_for_found_val]
+
+
+def marshall_json(resources: [], paths=[], flatten=False):
```

Comment:
I can add a test. The function was written because I was, at one point, separately inspecting svc and pod objects. We may do this again, the function is written. No need to move backwards.